circuits/src/zk_circuits/valid_wallet_update.rs

```diff
@@ -0,0 +1,1242 @@
+            let valid_mint = OrGate::multi_or(&[valid_mint1, valid_mint2, valid_mint3], cs);
+            cs.constrain(Variable::One() - valid_mint);
+        }
+
+        // Validate that the external transfer's mint did show up in exactly one of the balances
+        let transfer_applied = EqGadget::eq(external_transfer_applied, Variable::One().into(), cs);
+        let transfer_applied_or_zero = OrGate::or(transfer_applied, external_transfer_zero, cs);
+        cs.constrain(transfer_applied_or_zero - Variable::One());
+    }
+
+    /// Constrains all balance mints to be unique or zero
+    fn constrain_unique_balance_mints<CS: RandomizableConstraintSystem>(
+        wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        cs: &mut CS,
+    ) {
+        for i in 0..wallet.balances.len() {
+            for j in (i + 1)..wallet.balances.len() {
+                // Check whether balance[i] != balance[j]
+                let ij_unique = NotEqualGadget::not_equal(
+                    wallet.balances[i].mint.clone(),
+                    wallet.balances[j].mint.clone(),
+                    cs,
+                );
+
+                // Evaluate the polynomial mint * (1 - ij_unique) which is 0 iff
+                // the mint is zero, or balance[i] != balance[j]
+                let (_, _, constraint_poly) =
+                    cs.multiply(wallet.balances[i].mint.clone(), Variable::One() - ij_unique);
+                cs.constrain(constraint_poly.into());
+            }
+        }
+    }
+
+    // ----------
+    // | Orders |
+    // ----------
+
+    /// Validates the orders of the new wallet
+    fn validate_order_updates<CS: RandomizableConstraintSystem>(
+        old_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        new_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        new_timestamp: Variable,
+        cs: &mut CS,
+    ) {
+        // Ensure that all order's assert pairs are unique
+        Self::constrain_unique_order_pairs(new_wallet, cs);
+
+        // Ensure that the timestamps for all orders are properly set
+        Self::constrain_updated_order_timestamps(old_wallet, new_wallet, new_timestamp, cs);
+    }
+
+    /// Constrain the timestamps to be properly updated
+    /// For each order, if the order is unchanged from the previous wallet, the timestamp
+    /// should remain constant. Otherwise, the timestamp should be updated to the current
+    /// timestamp passed as a public variable
+    fn constrain_updated_order_timestamps<CS: RandomizableConstraintSystem>(
+        old_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        new_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        new_timestamp: Variable,
+        cs: &mut CS,
+    ) {
+        for (old_order, new_order) in old_wallet.orders.iter().zip(new_wallet.orders.iter()) {
```

Comment:
similar question about iterator ordering here, how can we be sure that the "logically same" orders are being zipped together here?
e.g. let's say we had an array of orders represented by asset pairs in the following way:
```
[
  (a0, a1),
  (a2, a3),
  (a4, a5),
]
```
and then we remove the second order `(a2, a3)`. is it the prover's responsibility to make sure they encode the new wallet with the following orders:
```
[
  (a0, a1),
  (zero_asset, zero_asset),
  (a4, a5)
]
```
so it zips properly?

Reply:
yes, prover's responsibility. We essentially define a canonical way of comparing two lists of balances or orders in the circuit, and it's the prover's responsibility to format the data into the witness such that this verifies.